web/src/immutant/web/internal/ring.clj

```diff
@@ -169,3 +172,28 @@
             :last-modified (-> vfile
                              .getPhysicalFile
                              ring.util.io/last-modified-date)})))))
+
+;;
+;; V1 comment:
+;; We don't use .getPathInfo (HttpServletRequest) or .getRequestPath
+;; (HttpServerExchange) since it is decoded. See IMMUTANT-195.
+;; ---
+;; This function ported from V1.x, also see IMMUTANT-610
+;;
+;; This function is called 'path-info-from-req' because it appears
+;; that naming it just 'path-info' breaks things, presumably because
+;; somehow there is an interaction with the path-info protocol
+;; function present in immutant.web.internal.servlet and
+;; immutant.web.internal.undertow.  In any case, changing the name of
+;; this function seemed to fix some bizarre exception throwing
+;; behavior in the integration tests.
+;;
+(defn path-info-from-req
+  "Takes a HttpServletRequest or HttpServerExchange and returns a
+  path-info string without any URL decoding performed upon it."
+  [request]
+  (let [path-info (subs (.getRequestURI request)
```

Comment:
This should probably be just inlined in the `undertow` and `servlet` namespaces, even if the code will be almost exactly the same. If we share a function between them, we can't type-hint the request (since it can either be an `HttpServerExchange` or `HttpServletRequest`), and without the type hint, reflection would kick in, which is relatively slow.